### ISSUE TITLE
combinator service unification

### ISSUE THREAD

**Issue by hawkw** [tower-rs/tower#498]
Currently, the `ServiceExt` trait has `with` and `try_with` methods for composing a `Service` with functions that modify the request type, and `map_err` and `map_ok` methods for composing a service with functions that modify the response type. Meanwhile, `ServiceBuilder` has `map_request` for composing a service with a request mapping function, and `map_response` for composing a service with a response-mapping function. These combinators are very similar in purpose, but are implemented using different middleware types that essentially duplicate the same behavior.

Before releasing 0.4, we should probably unify these APIs. In particular, it would be good to de-duplicate the middleware service types, and to unify the naming. 

Personally, I think the `ServiceExt` versions of `map_X` have a slightly better design, since they take `FnOnce + Clone`, rather than taking a `Fn` and `Arc`ing it internally. This avoids the overhead of `Arc` for mapping functions that are trivially clonable (e.g. function pointers)  but still allows the user to use `Arc`s when needed.

I'm open to bikeshedding what naming to settle on. `with` has an analogy with the `futures` crate's `Sink` methods, but I personally find `map_request` and `map_response` a *bit* clearer in the `tower` context. However, if we want to have both `Err` and `Ok` versions of these combinators, which I think we do, `map_response_err` and `map_response_ok` gets a bit wordy...perhaps we could just have `map_response` and `map_err`?

**Comment by hawkw** [comment]
cc @LucioFranco @hlb8122

**Comment by seanmonstar** [comment]
> `map_response_err` and `map_response_ok`

Drive-by: I prefer `map_request`, `map_response`, and `map_err`. 🏎️

**Comment by LucioFranco** [comment]
Drive-by: I agree with sean

**Comment by hlb8122** [comment]
`map_request` and `map_err` make sense and are inline with `map_ok` and `map_err` from `TryFutureExt`.
